File: backend/stages/stage1_rollup_wow.py
"""
Stage 1 — Roll-up WoW Z-Score
Roll up to (decline_code × channel), then compare current window's decline rate
to the SAME window (same day-of-week, same hour) across the prior 4 weeks.
Catches high-volume rate spikes that are statistically significant at roll-up level.
"""
import time
import pandas as pd
import numpy as np

ROLLUP_DIMS = ['decline_code', 'channel']


def run(df: pd.DataFrame, current_day: int = 0, current_hour: int = 14, threshold: float = 3.0) -> dict:
    start = time.perf_counter()

    curr = df[(df['week'] == 0) & (df['day_of_week'] == current_day) & (df['hour'] == current_hour)]
    hist = df[(df['week'] > 0) & (df['day_of_week'] == current_day) & (df['hour'] == current_hour)]

    # Roll up to (decline_code, channel)
    def rollup(frame, extra_group=None):
        groups = ([extra_group] if extra_group else []) + ROLLUP_DIMS
        g = frame.groupby(groups).agg(total=('total_count', 'sum'), declines=('decline_count', 'sum')).reset_index()
        g['rate'] = g['declines'] / g['total'].clip(lower=1)
        return g

    curr_r = rollup(curr)
    hist_r = rollup(hist, extra_group='week')

    # Historical stats per roll-up cell
    stats = hist_r.groupby(ROLLUP_DIMS).agg(
        wow_mean=('rate', 'mean'),
        wow_std=('rate', 'std'),
        n_weeks=('week', 'count'),
    ).reset_index()
    stats['wow_std'] = stats['wow_std'].fillna(0)

    merged = curr_r.merge(stats, on=ROLLUP_DIMS, how='left')
    merged['is_new'] = merged['wow_mean'].isna()
    merged['wow_mean'] = merged['wow_mean'].fillna(0)
    merged['z_score'] = (merged['rate'] - merged['wow_mean']) / merged['wow_std'].clip(lower=0.0005)
    merged.loc[merged['is_new'], 'z_score'] = 99.0
    merged['alerted'] = (merged['z_score'].abs() >= threshold) | merged['is_new']

    alerts = merged[merged['alerted']].sort_values('z_score', ascending=False)

    # Chart data: all monitors sorted by z-score (for bar chart)
    chart = merged.sort_values('z_score', ascending=False).head(30)
    chart_data = [
        {
            'label': f"{r['decline_code']}×{r['channel']}",
            'z_score': round(float(r['z_score']), 2),
            'rate': round(float(r['rate']), 4),
            'wow_mean': round(float(r['wow_mean']), 4),
            'alerted': bool(r['alerted']),
        }
        for _, r in chart.iterrows()
    ]

    exec_ms = (time.perf_counter() - start) * 1000
    return {
        'stage': 1,
        'name': 'Roll-up WoW Z-Score',
        'description': (
            f'Rolled to ({", ".join(ROLLUP_DIMS)}). '
            f'Current window vs same Mon 14:00 in weeks −1 through −4.'
        ),
        'n_monitors': int(len(merged)),
        'n_alerts': int(len(alerts)),
        'threshold': threshold,
        'execution_ms': round(exec_ms, 2),
        'chart_data': chart_data,
        'alerts': _to_records(alerts[ROLLUP_DIMS + ['total', 'declines', 'rate', 'wow_mean', 'wow_std', 'z_score', 'alerted']]),
    }


def _to_records(df: pd.DataFrame) -> list:
    records = []
    for _, r in df.iterrows():
        d = {}
        for k, v in r.items():
            if isinstance(v, (np.integer,)): d[k] = int(v)
            elif isinstance(v, (np.floating,)): d[k] = round(float(v), 4)
            elif isinstance(v, (np.bool_,)): d[k] = bool(v)
            else: d[k] = v
        records.append(d)
    return records

File: backend-java/src/main/java/com/anomalydetection/payment/Stage1RollupWow.java
package com.anomalydetection.payment;

import java.util.*;

import static com.anomalydetection.payment.RowAggregation.*;

/** Port of stages/stage1_rollup_wow.py.
 * Two queries -- current window and prior weeks, both grouped by
 * (decline_code, channel) -- then a small downstream z-score per monitor. */
public class Stage1RollupWow {

    record RollupKey(String declineCode, String channel) {
        static RollupKey of(PaymentRow r) { return new RollupKey(r.declineCode, r.channel); }
    }

    public static Map<String, Object> run(List<PaymentRow> df, int currentDay, int currentHour, double threshold) {
        long startNs = System.nanoTime();

        // Query 1 -- current window, rolled up to (decline_code, channel)
        Map<RollupKey, Agg> current = groupByKey(currentWindow(df, currentDay, currentHour), RollupKey::of);

        // Query 2 -- prior 4 weeks, same slot, per (week, decline_code, channel)
        Map<Integer, Map<RollupKey, Agg>> historical = groupByWeekThenKey(historicalWindow(df, currentDay, currentHour), RollupKey::of);

        // Downstream: transpose to per-monitor list of historical rates
        Map<RollupKey, List<Double>> historicalRates = new LinkedHashMap<>();
        historical.forEach((week, byKey) -> byKey.forEach((key, agg) ->
                historicalRates.computeIfAbsent(key, k -> new ArrayList<>()).add(agg.rate())));

        List<Map<String, Object>> merged = new ArrayList<>();
        for (Map.Entry<RollupKey, Agg> e : current.entrySet()) {
            RollupKey key = e.getKey();
            Agg agg = e.getValue();
            double rate = agg.rate();
            List<Double> history = historicalRates.getOrDefault(key, List.of());
            boolean isNew = history.isEmpty();
            double z = WowMath.zScore(rate, history, 0.0005);
            boolean alerted = Math.abs(z) >= threshold || isNew;

            Map<String, Object> m = new LinkedHashMap<>();
            m.put("decline_code", key.declineCode());
            m.put("channel", key.channel());
            m.put("total", agg.total);
            m.put("declines", agg.declines);
            m.put("rate", rate);
            m.put("wow_mean", isNew ? 0 : WowMath.mean(history));
            m.put("wow_std", isNew ? 0 : WowMath.std(history));
            m.put("z_score", z);
            m.put("is_new", isNew);
            m.put("alerted", alerted);
            merged.add(m);
        }

        merged.sort((x, y) -> Double.compare((Double) y.get("z_score"), (Double) x.get("z_score")));
        List<Map<String, Object>> alerts = merged.stream().filter(m -> (Boolean) m.get("alerted")).toList();

        List<Map<String, Object>> chartData = new ArrayList<>();
        for (Map<String, Object> r : merged.subList(0, Math.min(30, merged.size()))) {
            Map<String, Object> c = new LinkedHashMap<>();
            c.put("label", r.get("decline_code") + "×" + r.get("channel"));
            c.put("z_score", WowMath.round((Double) r.get("z_score"), 2));
            c.put("rate", WowMath.round((Double) r.get("rate"), 4));
            c.put("wow_mean", WowMath.round((Double) r.get("wow_mean"), 4));
            c.put("alerted", r.get("alerted"));
            chartData.add(c);
        }

        List<Map<String, Object>> alertRecords = new ArrayList<>();
        for (Map<String, Object> r : alerts) {
            Map<String, Object> d = new LinkedHashMap<>();
            d.put("decline_code", r.get("decline_code"));
            d.put("channel", r.get("channel"));
            d.put("total", ((Long) r.get("total")).intValue());
            d.put("declines", ((Long) r.get("declines")).intValue());
            d.put("rate", WowMath.round((Double) r.get("rate"), 4));
            d.put("wow_mean", WowMath.round((Double) r.get("wow_mean"), 4));
            d.put("wow_std", WowMath.round((Double) r.get("wow_std"), 4));
            d.put("z_score", WowMath.round((Double) r.get("z_score"), 4));
            d.put("alerted", r.get("alerted"));
            alertRecords.add(d);
        }

        double execMs = (System.nanoTime() - startNs) / 1e6;

        Map<String, Object> result = new LinkedHashMap<>();
        result.put("stage", 1);
        result.put("name", "Roll-up WoW Z-Score");
        result.put("description", "Rolled to (decline_code, channel). Current window vs same Mon 14:00 in weeks −1 through −4.");
        result.put("n_monitors", merged.size());
        result.put("n_alerts", alerts.size());
        result.put("threshold", threshold);
        result.put("execution_ms", WowMath.round(execMs, 2));
        result.put("chart_data", chartData);
        result.put("alerts", alertRecords);
        return result;
    }
}
